libs/langchain-community/src/vectorstores/elasticsearch.ts

```diff
@@ -266,16 +270,23 @@ export class ElasticVectorSearch extends VectorStore {
       mappings: {
         dynamic_templates: [
           {
-            // map all metadata properties to be keyword
-            "metadata.*": {
+            // map all metadata properties to be keyword except loc
+            metadata_except_loc: {
```

Comment:
Is there we can generalize this a bit more to handle all object properties?

Reply:
Thanks for reviewing!

One reason I did it this way is that I tried to be as focused as possible on fixing the bug and leaving other existing behavior alone.

We could possibly use object type for every metadata field and perhaps it's the right thing to do given that `metadata` is defined as `Record<string, any>`. I think it would be like this:

```js
            "metadata.*": {
              match_mapping_type: "*",
              match: "metadata.*",
              mapping: { type: "object" },
            },
```

But I think there are a couple of reasonable arguments for keeping `metadata` fields declared as `keyword` in Elasticsearch. Searching through the codebase, I see that `loc` is the only field that is ever anything but a flat number or string. And I think being able to filter by `keyword` type fields is much simpler and more performant in Elasticsearch. For my use case, for example, I want to augment an embedding based search with keyword filtering, so I'd prefer to keep them as keywords.

Another solution to #2857 could be to change the way `loc` is stored so that it can be stored in a keyword field. The `TextSplitter`s could store something like `line_from` and `line_to`. In that case, it might make sense to update the type of `metadata` as well. This might be a bigger lift since it would potentially affect anything downstream of those `TextSplitter`s.

Reply:
Yeah we can't change text splitter metadata yet, though I think flattening it would be nice when we're ready to make a breaking change.

Reply:
I'm ok with this but can we add a TODO in the code as well? Also remember to run `yarn lint`!